src/vsr/replica.zig

```diff
@@ -1665,7 +1665,25 @@ pub fn ReplicaType(
 
             for (view_headers.slice) |*header| {
                 if (header.op <= self.op_checkpoint_trigger()) {
+                    // Consider this case:
+                    // 1. Replica A prepares and commits op X (as backup).
+                    // 2. Replica A crashes.
+                    // 3. Prepare X is corrupted in the WAL.
+                    // 4. Replica A recovers into status=recovering_head.
+                    // 5. Replica A receives a start_view, which includes the header of X.
+                    // 6. Replica A participates in a view-change.
+                    // During this view-change, replica A *must not* nack op X.
+                    // Otherwise, op X might be truncated from the new view.
+                    // To ensure this, leave the entry as faulty.
```

Comment:
Hm, actually, shouldn't

```
            if (journal.has(header)) {
                assert(journal.dirty.bit(slot));
                // Do not clear any faulty bit for the same entry.
            } else {
```

logic in Journal already handle both of these cases?

Reply:
To answer your second question — that journal logic doesn't always help.
In the example I put in the code, we have no header at all (`command=reserved`) prior to calling `replace_header` (the WAL was corrupt).

For your former question:
We could put this same logic in `primary_set_log_from_do_view_change_messages`, but I don't _think_ it is necessary.
The difference is that the headers of this new view cannot be lost, even if we crash + recover the WAL — because they are stored in our superblock's `vsr_headers`, and restored (if necessary) during `open`.

Reply:
:thinking: tbh, I am not a fan of this save-restore logic, especially given that it papers over similar logic on journal. 

Could we do something like

```zig
            if (journal.has(header) or journal.faulty(slot) and journal.headers[slot].command == .reserved) {
                assert(journal.dirty.bit(slot));
                // Do not clear any faulty bit for potentially the same entry.
```

in the journal perhaps?

Reply:
I need to play with this a bit, I think there are some tricky cases.
I think we need to leave `set_header_as_dirty` as-is since leaving faulty intact in all cases could impact availability.
But we could add a new function that does leave faulty intact, which the replica can use in some circumstances. (Hand-waiving here b/c I'm not sure which all circumstances those are yet...)

Also I think we can't check `command == reserved` because `Replica.open` loads headers from vsr_headers but leaves `faulty` intact.